cpp/src/arrow/filesystem/gcsfs.h

```diff
@@ -27,11 +27,37 @@
 namespace arrow {
 namespace fs {
 
-struct GcsCredentials;
+// Opaque wrapper for GCS's library credentials to avoid exposing in Arrow headers.
+struct GcsCredentialsHolder;
+class GcsFileSystem;
+
+/// \brief Container for GCS Credentials an information necessary to recreate
+///  them.
+class GcsCredentials {
+ public:
+  bool Equals(const GcsCredentials& other) const;
+  bool anonymous() const { return anonymous_; }
+  const std::string& access_token() { return access_token_; }
+  TimePoint expiration() const { return expiration_; }
+  const std::string& target_service_account() { return target_service_account_; }
+  const std::string& json_credentials() { return json_credentials_; }
+  const std::shared_ptr<GcsCredentialsHolder>& holder() const { return holder_; }
```

Comment:
Please make sure all accessors are this-const.

Reply:
done.